### DISCUSSION TITLE
Using redirect or router outside of components

### DISCUSSION THREAD

**Discussion by Amerehei** [vercel/next.js#56678] [App Router]
I have tried https://github.com/vercel/next.js/discussions/17046 but it seems it's out of date and in Nextj.s 13 and AppRouter we need some other solutions

```
router.js:98 Uncaught (in promise) Error: No router instance found.
You should only use "next/router" on the client side of your app.
```

**Comment by SidKH** [comment]
What are you trying to achieve?

**Comment by Amerehei** [comment]
I have some codes outside of React components, they are listening to some events and emit some other events. In response of one of events I would like to redirect page to another page but I have not access to router and redirect doesn't work.

I can pass router from a component to the non-component code, but I think there are better approaches

**Comment by SidKH** [comment]
Try 👇

```js
import { redirect } from "next/navigation";

...
redirect("/my-path");
...
```

**Comment by Amerehei** [comment]
As I said before `redirect` and `router` don't work

```
redirect.js:47 Uncaught (in promise) Error: NEXT_REDIRECT
```

**Comment by SidKH** [comment]
I see, I don't think it's possible atm.  
  
The `redirect` can only be called during rendering and there is no global router instance exported from `next/navigation` (App Router)

> I can pass router from a component to the non-component code

You can create a custom hook that automatically injects the router instance into your event emitter
https://codesandbox.io/p/sandbox/little-water-twh6ps?file=%2Fapp%2Fpage.tsx%3A9%2C1
